In [1]:
import pickle as pkl 

import pandas as pd

from sklearn.preprocessing import StandardScaler

from sklearn.cluster import KMeans

In [2]:
with open('artist_features.pkl', 'rb') as f:
    artist_features = pkl.load(f)

In [4]:
# focus on rates to start rather than song length
X = artist_features.drop(columns=['lines', 'words', 'syllables'])
X

Lyrical/Misc  Bragging/Combative  Money/Party  Romance/Sex  \
artist                                                                   
03 Greedo       0.012388            0.040464     0.021360     0.011842   
070 Shake       0.015179            0.001804     0.021861     0.036665   
2 Chainz        0.018329            0.023071     0.021967     0.006407   
2 Pistols       0.013755            0.030431     0.018532     0.021855   
21 Savage       0.007464            0.046448     0.046098     0.008801   
...                  ...                 ...          ...          ...   
Zebra Katz      0.025077            0.012421     0.018921     0.016527   
Zelooperz       0.018682            0.037817     0.003213     0.006282   
Zion I          0.031290            0.000425     0.003131     0.006091   
bbno$           0.009183            0.010923     0.004047     0.009563   
will.i.am       0.004096            0.019755     0.047252     0.040928   

            words_per_line  unique_words  unique_word_rate  \
artist                                                       
03 Greedo        35.125816    185.375000          0.070810   
070 Shake        30.507007    137.470588          0.071274   
2 Chainz         32.916186    194.111111          0.090778   
2 Pistols        28.376804    178.272727          0.085574   
21 Savage        33.424155    189.095238          0.070440   
...                    ...           ...               ...   
Zebra Katz       25.122420    132.944444          0.087163   
Zelooperz        29.884914    151.060606          0.095839   
Zion I           34.300381    210.200000          0.104322   
bbno$            31.152735     92.333333          0.055000   
will.i.am        33.323155    163.733333          0.067369   

            syllables_per_line  syllables_per_word  
artist                                              
03 Greedo             9.211347            0.261876  
070 Shake             8.091570            0.264949  
2 Chainz              8.641371            0.262462  
2 Pistols             7.243027            0.255224  
21 Savage             8.484020            0.253769  
...                        ...                 ...  
Zebra Katz            6.521429            0.259716  
Zelooperz             7.746176            0.258715  
Zion I                8.986945            0.262015  
bbno$                 8.575163            0.275106  
will.i.am             8.664136            0.259974  

[1260 rows x 9 columns]

In [9]:
scaler = StandardScaler()
X_scaled = pd.DataFrame(scaler.fit_transform(X), columns=X.columns, index=X.index)

In [10]:
X_scaled

Lyrical/Misc  Bragging/Combative  Money/Party  Romance/Sex  \
artist                                                                   
03 Greedo      -1.108094            1.115106     0.780540    -0.227123   
070 Shake      -0.834664           -1.035582     0.819309     1.803736   
2 Chainz       -0.526146            0.147511     0.827469    -0.671790   
2 Pistols      -0.974242            0.556940     0.561764     0.592036   
21 Savage      -1.590426            1.447984     2.694137    -0.475869   
...                  ...                 ...          ...          ...   
Zebra Katz      0.134856           -0.444954     0.591835     0.156195   
Zelooperz      -0.491603            0.967847    -0.623256    -0.681979   
Zion I          0.743504           -1.112303    -0.629578    -0.697615   
bbno$          -1.422031           -0.528276    -0.558746    -0.413536   
will.i.am      -1.920378           -0.036947     2.783435     2.152477   

            words_per_line  unique_words  unique_word_rate  \
artist                                                       
03 Greedo         0.514503     -0.344285         -1.123740   
070 Shake        -0.442401     -1.017869         -1.090486   
2 Chainz          0.056722     -0.221446          0.307881   
2 Pistols        -0.883727     -0.444150         -0.065240   
21 Savage         0.161960     -0.291975         -1.150268   
...                    ...           ...               ...   
Zebra Katz       -1.557956     -1.081511          0.048671   
Zelooperz        -0.571284     -0.826780          0.670745   
Zion I            0.343493      0.004780          1.278952   
bbno$            -0.308622     -1.652544         -2.257284   
will.i.am         0.141036     -0.648589         -1.370463   

            syllables_per_line  syllables_per_word  
artist                                              
03 Greedo             0.482583           -0.244783  
070 Shake            -0.394098            0.285492  
2 Chainz              0.036345           -0.143757  
2 Pistols            -1.058428           -1.392879  
21 Savage            -0.086846           -1.644092  
...                        ...                 ...  
Zebra Katz           -1.623372           -0.617658  
Zelooperz            -0.664509           -0.790359  
Zion I                0.306897           -0.220902  
bbno$                -0.015490            2.038551  
will.i.am             0.054167           -0.573045  

[1260 rows x 9 columns]

In [11]:
clusters = 5
km = KMeans(n_clusters=clusters, random_state=77, n_init=3)
km.fit(X_scaled)

KMeans(n_clusters=5, n_init=3, random_state=77)

In [16]:
X_labeled = X.copy()
X_labeled['cluster'] = km.labels_
X_labeled

Lyrical/Misc  Bragging/Combative  Money/Party  Romance/Sex  \
artist                                                                   
03 Greedo       0.012388            0.040464     0.021360     0.011842   
070 Shake       0.015179            0.001804     0.021861     0.036665   
2 Chainz        0.018329            0.023071     0.021967     0.006407   
2 Pistols       0.013755            0.030431     0.018532     0.021855   
21 Savage       0.007464            0.046448     0.046098     0.008801   
...                  ...                 ...          ...          ...   
Zebra Katz      0.025077            0.012421     0.018921     0.016527   
Zelooperz       0.018682            0.037817     0.003213     0.006282   
Zion I          0.031290            0.000425     0.003131     0.006091   
bbno$           0.009183            0.010923     0.004047     0.009563   
will.i.am       0.004096            0.019755     0.047252     0.040928   

            words_per_line  unique_words  unique_word_rate  \
artist                                                       
03 Greedo        35.125816    185.375000          0.070810   
070 Shake        30.507007    137.470588          0.071274   
2 Chainz         32.916186    194.111111          0.090778   
2 Pistols        28.376804    178.272727          0.085574   
21 Savage        33.424155    189.095238          0.070440   
...                    ...           ...               ...   
Zebra Katz       25.122420    132.944444          0.087163   
Zelooperz        29.884914    151.060606          0.095839   
Zion I           34.300381    210.200000          0.104322   
bbno$            31.152735     92.333333          0.055000   
will.i.am        33.323155    163.733333          0.067369   

            syllables_per_line  syllables_per_word  cluster  
artist                                                       
03 Greedo             9.211347            0.261876        2  
070 Shake             8.091570            0.264949        4  
2 Chainz              8.641371            0.262462        2  
2 Pistols             7.243027            0.255224        2  
21 Savage             8.484020            0.253769        0  
...                        ...                 ...      ...  
Zebra Katz            6.521429            0.259716        4  
Zelooperz             7.746176            0.258715        2  
Zion I                8.986945            0.262015        3  
bbno$                 8.575163            0.275106        4  
will.i.am             8.664136            0.259974        0  

[1260 rows x 10 columns]

In [41]:
X_labeled.groupby('cluster').count()

Lyrical/Misc  Bragging/Combative  Money/Party  Romance/Sex  \
cluster                                                               
0                 122                 122          122          122   
1                 217                 217          217          217   
2                 277                 277          277          277   
3                 401                 401          401          401   
4                 243                 243          243          243   

         words_per_line  unique_words  unique_word_rate  syllables_per_line  \
cluster                                                                       
0                   122           122               122                 122   
1                   217           217               217                 217   
2                   277           277               277                 277   
3                   401           401               401                 401   
4                   243           243               243                 243   

         syllables_per_word  
cluster                      
0                       122  
1                       217  
2                       277  
3                       401  
4                       243

In [40]:
list(X_labeled.reset_index()[X_labeled.reset_index()['cluster'] == 4]['artist'])

['070 Shake',
 '24kGoldn',
 '360',
 'A+',
 'A-Q',
 'A.CHAL',
 'A.L.T.',
 'Afrika Bambaataa',
 'Afroman',
 'Akon',
 'Aminé',
 'Amir Obè',
 'Anthony Parker',
 'BJ the Chicago Kid',
 'Baby Bash',
 'Baeza',
 'Bas',
 'Benny Blanco',
 'Bia',
 'Big Boi',
 'Big Shaq',
 'Blac Youngsta',
 'Black Milk',
 'Blu',
 'Bobby Brackins',
 'Bohemia',
 'Boity',
 'Boots Riley',
 'Burna Boy',
 'Butch Cassidy',
 'CL',
 'Cage',
 'Cassidy',
 'Cazwell',
 'CeeLo Green',
 'Channel 7',
 'Chris Brown',
 'Consequence',
 'Cool C',
 'Curtiss King',
 "D'Angelo",
 'DJ Head',
 'DJ Premier',
 'DJ Run',
 'DJ Shadow',
 'DLOW',
 'Danny Boy',
 'David Rush',
 'Desiigner',
 'Detail',
 'Deuce',
 'Diamond D',
 'Disco D',
 'Doja Cat',
 'Dominic Fike',
 'Dominique Young Unique',
 'Domino',
 'Drag-On',
 'Duncan Mighty',
 'Elephant Man',
 'Elucid',
 'Emtee',
 'Eric Stanley',
 'Fatboi',
 'Father MC',
 'Fatman Scoop',
 'Fergie',
 'Flavor Flav',
 'Flo Rida',
 'Focus...',
 'Frank Ocean',
 'Fred Durst',
 'Freekey Zekey',
 'Frost',
 'G-Eazy

#### Notes: 
#### k=5:
Cluster 0 has 122, Cluster 3 has 401, rest around 2-300
* 0: Count: 122 -- Money/Party dominated, average words/line, 2nd lowest unique word usage, average syllables per line, lowest syllables per word 
* 1: Count: 217 -- Lyrical/Misc dominated, mild Bragging/Combative, highest words/line, highest unique words/rate, highest syllables per line and word
* 2: Count: 277 -- Bragging/Combative Dominated, average everything
* 3: Count: 401 -- Lyrical/Misc dominated, mild Bragging/Combative AND Romance/Sex, average words/line, above average unique word rate, average syllables per line/word
* 4: Count: 243 -- Romance/Sex dominated, mild Lyrical/Misc, lowest words/line by far, lowest unique words/rate, lowest syllables/line, but average per word

Example artists in Each Cluster:
* 0: Cardi B/Nicki Minaj/Megan Thee Stallion, Roddy Ricch, Travis Scott, Young Thug, XXXTentacion
* 1: Aesop Rock, Eminem, Ghostface Killah, J. Cole, Lil Wayne
* 2: 2 Chainz, 50 Cent, DMX, Snoop Dogg, Wiz Khalifa
* 3: Andre 3000, Chance the Rapper, Childish Gambino, Kendrick Lamar, Lupe Fiasco, Macklemore
* 4: Chris Brown, Jason Derulo, Kid Cudi, Lauryn Hill, Post Malone, The Weeknd

Cluster Analysis: 
* 0: Hardto interpret. Chiefly generic money/party semantic content, most simplistic word usage. Seems to house most of the trap/'vibe'/mumble rappers. Some artists don't seem very accurately placed (e.g. Cardi/Nicki/Megan). 
* 1: Clearly home to most 'lyrical' rappers. Dominant cluster in terms of vocabulary and syllable density.
* 2: Very non-descript, middle of the pack by every metric. Seems to house a lot of older, 'standard' flow rappers.
* 3: Seems to capture most 'musically' oriented rappers -- lyrical diversity but not massive density, widest variety of topics.
* 4: Effectively captures most R&B artists. Romance/Sex oriented lyrics with minimal lyrical diversity.